File: CollatzSequence/Collatz.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Globalization;
using System.Numerics;
using System.Text;

namespace CollatzSequence;

public class Collatz
{
	/// <summary>
	/// The starting value for generating the Collatz Sequence.
	/// </summary>
	private BigInteger _startingNumber;

	/// <summary>
	/// The Collatz Sequence.
	/// </summary>
	private ReadOnlyCollection<BigInteger> _collatzSequence;

	/// <summary>
	/// The getter and setter for _startingNumber.
	/// </summary>
	public BigInteger StartingNumber
	{
		get
		{
			return this._startingNumber;
		}

		set
		{
			if (value <= 0)
			{
				throw new ArgumentOutOfRangeException(nameof(value), "The starting number must be a positive integer.");
			}
			this._startingNumber = value;
			this._collatzSequence = GenerateSequence();
		}
	}

	/// <summary>
	/// The getter for the list of the Collatz Sequence.
	/// </summary>
	public ReadOnlyCollection<BigInteger> CollatzSequence
	{
		get
		{
			return this._collatzSequence;
		}
	}
		
	/// <summary>
	/// Constructor takes in the starting value for the sequence and generates the Collatz Sequence.
	/// Sets sequence to -1 if an overflow exception is thrown during sequence generation.
	/// </summary>
	/// <param name="number">The starting value for the sequence.</param>
	/// <exception cref="ArgumentOutOfRangeException">Exception if value <= 0 is passed.</exception>
	public Collatz(BigInteger number)
	{
		if (number <= 0)
		{
			throw new ArgumentOutOfRangeException(nameof(number), "The starting number must be a positive integer.");
		}

		// Set starting number from input.
		_startingNumber = number;


		// Set the complete sequence.
		_collatzSequence = GenerateSequence();
	}

	private ReadOnlyCollection<BigInteger> GenerateSequence()
	{
		// Initialize a new BigInteger with _startingNumber, to then modify.
		var currentNumber = _startingNumber;
		// Initialize a new list with the user number as the first element.
		var sequence = new List<BigInteger>() { currentNumber };

		// Due to the nature of BigInteger, and overflow exception may be thrown.
		try
		{
			// Loop until the sequence reaches the end value of one.
			while (currentNumber > 1)
			{
				// If even, divide the number by 2.
				// If odd, multiply the number by 3, add 1, and divide by 2.
				currentNumber = currentNumber % 2 == 0 ? currentNumber / 2 : (3 * currentNumber + 1) / 2;
				// Add the new number to the sequence list.
				sequence.Add(currentNumber);
			}
		}
		// Catch the overflow exception.
		catch (OverflowException)
		{
			// Clear the list and set the first element to -1 to indicate Collatz failure.
			sequence.Clear();
			sequence.Add(-1);
		}

		// Returns the Collatz Sequence.
		return sequence.AsReadOnly();
	}

	/// <summary>
	/// Builds a string with each number of the sequence on a line to display.
	/// </summary>
	/// <returns>The Collatz Sequence as a string to display.</returns>
	public string DisplaySequence()
	{
		// Build return string.
		var sb = new StringBuilder();
		foreach(var number in CollatzSequence)
		{
			sb.AppendLine(number.ToString("G", NumberFormatInfo.CurrentInfo));
		}

		// Returns the Collatz Sequence string to display.
		return sb.ToString();
	}
}

File: Typescript/src/collatz.ts
/**
 * Collatz class for generating a Collatz Sequence from a starting value.
 */
export default class Collatz {
    private starting_number: bigint;
    private collatz_sequence: Array<bigint>;

    /**
     * The constructor for the Collatz class. Sets starting_number to the passed
     * value and sets the Collatz Sequence for that number. The starting value
     * must be a positive integer.
     *
     * @param starting_number the starting number of the Collatz Sequence.
     * @throws if the starting number less than or equal to zero is passed.
     */
    public constructor(starting_number: bigint) {
        if (starting_number <= BigInt(0)) {
            throw new Error('The starting number must be a positive integer.');
        } else {
            this.starting_number = starting_number;
            this.collatz_sequence = this.generate_sequence();
        }
    }

    /**
     * The getter for starting_number.
     *
     * @return the starting number.
     */
    public get_starting_number(): bigint {
        return this.starting_number;
    }

    /**
     * The setter for the startingNumber of the Collatz Sequence.  When a new
     * starting number is set, the Collatz Sequence is automatically generated
     * for that number. The starting value must be a positive integer.
     *
     * @param starting_number the starting number of the Collatz Sequence.
     * @throws if the starting number less than or equal to zero is passed.
     */
    public set_starting_number(starting_number: bigint): void {
        if (starting_number <= BigInt(0)) {
            throw new Error('The starting number must be a positive integer.');
        } else {
            this.starting_number = starting_number;
            this.collatz_sequence = this.generate_sequence();
        }
    }

    /**
     * The getter for collatz_sequence.
     *
     * @return a readonly array.
     */
    public get_collatz_sequence(): readonly bigint[] {
        return Object.freeze(this.collatz_sequence);
    }

    /**
     * Returns the length of the sequence.
     *
     * @return the integer value of the length of the sequence.
     */
    public get_sequence_length(): number {
        return this.collatz_sequence.length;
    }

    /**
     * Builds a string with all the values in collatz_sequence, with each
     * value on its own line.
     *
     * @return a string of the Collatz Sequence.
     */
    public display_sequence(): string {
        return this.collatz_sequence.join('\n');
    }

    /**
     * Generates the Collatz Sequence from the starting number.
     *
     * @return a bigint array of the sequence.
     */
    private generate_sequence(): Array<bigint> {
        // Initialize current_number with the user number.
        let current_number: bigint = this.starting_number;
        // Initialize a new list with the user number as the first element.
        const sequence: Array<bigint> = [ current_number ];

        // Loop until the sequence reaches the end value of one.
        while (current_number !== BigInt(1)) {
            // If even, divide the number by 2.
            // If odd, multiply the number by 3, add 1, and divide by 2.
            current_number = current_number % BigInt(2) === BigInt(0) ?
                current_number / BigInt(2) :
                current_number * BigInt(3) + BigInt(1);
            // Add the new number to the sequence list.
            sequence.push(current_number);
        }

        // Set the Collatz Sequence.
        return sequence;
    }
}
